composer/utils/collect_env.py

```diff
@@ -0,0 +1,245 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+"""Helpers to gather system information for debugging and bug reporting.
+
+Leverages PyTorch's :mod:`torch.utils.collect_env` package to gather pertinent system information.
+The following information is additionally collected to faciliate Comopser specific debug:
+
+* Composer version
+* Number of nodes
+* Host processor model name
+* Host processor physical core count
+* Number of accelerators per node
+* Accelerator model name
+
+This package can be invoked as a standalone console script or can be invoked from within
+an application to gather and generate a system environment report.
+
+The module can be invoked by using the entrypoint alias:
+
+.. code-block::
+
+    composer_collect_env
+
+Or manually as a standalone script:
+
+.. code-block::
+
+    python composer/utils/collect_env.py
+
+To generate a system report from within a user application see :func:`print_env`.
+
+A custom excepthook wrapper is also provided which extends the default :func:`sys.excepthook`
+to automatically collect system information when an exception is raised.
+
+To override the default :func:`sys.excepthook` see :func:`configure_excepthook`.
+
+To restore the default :func:`sys.excepthook` see :func:`restore_excepthook`.
+"""
+
+import sys
+import time
+from typing import NamedTuple
+
+import cpuinfo
+import psutil
+
+__all__ = ["configure_excepthook", "print_env", "restore_excepthook"]
+
+# Check if PyTorch is installed
+try:
+    import torch.utils.collect_env as torchenv
+    from torch.cuda import get_device_name as accel_device_name
+    from torch.cuda import is_available as cuda_available
+    TORCH_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    TORCH_AVAILABLE = False
+
+# Check if Composer is installed
+try:
+    import composer
+    from composer.utils import dist
+    COMPOSER_AVAILABLE = True
+except (ImportError, NameError, AttributeError, OSError):
+    COMPOSER_AVAILABLE = False
+
+COMPOSER_OPEN_ISSUE_URL = "https://github.com/mosaicml/composer/issues/new/choose"
+EXCEPTION_MSG = f"If you believe this exception was raised due to a Composer bug, " + \
```

Comment:
Nit: "If you believe" is a bit ambiguous.

Generally speaking:

- If it's a `AssertionError, then it's definitely a Composer Bug. A bug report should always be filed.
- If it's a `RuntimeError, then it's likely a Composer bug, or someone using something very incorrectly.
- If it's a `ValueError`, `TypeError`, then it's likely a user bug, but could be a composer bug.
- If it's another exception type, then not sure -- could be anything?

Could we customize the message based off of this?

Reply:
Sure, changing the messaging based on the exception.